### PR TITLE
use local scheduler when PBS_NODEFILE is not available

### LINKED ISSUE #106: use local scheduler when PBS_NODEFILE is not available
right now it causes an error
```
$ mympirun hostname
2017-02-22 17:07:29,771 WARNING    mympirun.IntelHydraMPI_PBS MainThread  set_nodes: failed to get PBS_NODEFILE from environment
2017-02-22 17:07:29,772 ERROR      mympirun        MainThread  Main failed
Traceback (most recent call last):
  File "/vscmnt/gent_vulpix/_/user/data/gent/vsc413/vsc41353/mympirun-dev/software/vsc-mympirun/dev/lib/python2.7/site-packages/vsc_mympirun-3.4.4-py2.7.egg/EGG-INFO/scripts/mympirun", line 123, in main
    instance = getinstance(*instance_options)
  File "/user/data/gent/vsc413/vsc41353/mympirun-dev/software/vsc-mympirun/dev/lib/python2.7/site-packages/vsc_mympirun-3.4.4-py2.7.egg/vsc/mympirun/factory.py", line 66, in getinstance
    return coupler_class(options=options.options, cmdargs=options.args)
  File "/user/data/gent/vsc413/vsc41353/mympirun-dev/software/vsc-mympirun/dev/lib/python2.7/site-packages/vsc_mympirun-3.4.4-py2.7.egg/vsc/mympirun/factory.py", line 58, in __init__
    super(Coupler, self).__init__(**kwargs)
  File "/user/data/gent/vsc413/vsc41353/mympirun-dev/software/vsc-mympirun/dev/lib/python2.7/site-packages/vsc_mympirun-3.4.4-py2.7.egg/vsc/mympirun/mpi/mpi.py", line 310, in __init__
    super(MPI, self).__init__(**kwargs)
  File "/user/data/gent/vsc413/vsc41353/mympirun-dev/software/vsc-mympirun/dev/lib/python2.7/site-packages/vsc_mympirun-3.4.4-py2.7.egg/vsc/mympirun/rm/sched.py", line 107, in __init__
    self.set_nodes()
  File "/vscmnt/gent_vulpix/_/user/data/gent/vsc413/vsc41353/mympirun-dev/software/vsc-mympirun/dev/lib/python2.7/site-packages/vsc_mympirun-3.4.4-py2.7.egg/vsc/mympirun/rm/pbs.py", line 49, in set_nodes
    self.log.raiseException("set_nodes: failed to get %s from environment"  % nodevar)
  File "/usr/lib/python2.7/site-packages/vsc/utils/fancylogger.py", line 329, in raiseException
    raise exception, message, tb
Exception: set_nodes: failed to get PBS_NODEFILE from environment
```

### PR DESCRIPTION
**Pull request by Caylo**
fixes #106

### CODE REVIEW

File: lib/vsc/mympirun/rm/sched.py [hunk 1]
```diff
@@ -49,6 +49,14 @@ def what_sched(requested):
                 return sched, found_sched
         LOGGER.warn("%s scheduler was requested, but mympirun failed to find an implementation", requested)
 
+    # if PBS_NODEFILE is not present, use local scheduler
+    filename = os.environ.get('PBS_NODEFILE', None)
+    if filename is None:
```
Comment (boegel) [inline lib/vsc/mympirun/rm/sched.py]: only used in one place, so no need for a variable

also, `os.environ.get` uses `None` as fallback value by default, no need to specify it

and usually `os.getenv` is used instead (same behaviour)

```python
if not os.getenv('PBS_NODEFILE'):
```

File: lib/vsc/mympirun/rm/sched.py [hunk 2]
```diff
@@ -49,6 +49,14 @@ def what_sched(requested):
...
+        for sched in found_sched:
+            if sched._is_sched_for("local"):
+                LOGGER.debug("PBS_NODEFILE not found, using local scheduler")
+                return sched, found_sched
```
Comment (boegel) [inline lib/vsc/mympirun/rm/sched.py]: this same loop is also used below, please don't duplicate code but factor it out in a function you call out to

File: lib/vsc/mympirun/rm/sched.py (hunk 1 above)
Comment (boegel) [inline lib/vsc/mympirun/rm/sched.py]: also, we should only be doing this if no scheduler was specified (or if the default `pbs` scheduler was used)?

File: test/end2end.py [hunk 3]
```diff
@@ -308,3 +309,14 @@ def change_env(self, cores):
         pbsnodefile.write('\n'.join(['localhost'] * cores))
         pbsnodefile.close()
         os.environ['PBS_NODEFILE'] = pbsnodefile.name
+
+
+    def test_unset_nodefile(self):
+        self.assertTrue(schedm.what_sched(False)[0].__name__ == 'PBS')
```
Comment (boegel) [inline test/end2end.py]: test with `isinstance(schedm.what_sched(False), PBS)` ?

File: test/end2end.py [hunk 4]
```diff
@@ -308,3 +309,14 @@ def change_env(self, cores):
...
+        nodefile = os.environ['PBS_NODEFILE']
+        del os.environ['PBS_NODEFILE']
+        # fall back to local if PBS_NODEFILE is not available
+        self.assertTrue(schedm.what_sched(False)[0].__name__ == 'Local')
```
Comment (boegel) [inline test/end2end.py]: test with `isinstance`?

File: test/end2end.py [hunk 5]
```diff
@@ -308,3 +309,14 @@ def change_env(self, cores):
...
+
+        # reset nodefile for proper test cleanup
+        os.environ['PBS_NODEFILE'] = nodefile
```
Comment (boegel) [inline test/end2end.py]: I don't think this is needed, the tests already restore the environment via `tearDown`? do double check...

File: lib/vsc/mympirun/rm/sched.py [hunk 6]
```diff
@@ -42,27 +42,36 @@ def what_sched(requested):
...
-        LOGGER.debug("No scheduler found in environment, trying local")
-        if sched._is_sched_for("local"):
-            return sched, found_sched
+    LOGGER.debug("No scheduler found in environment, trying local")
+    return local_sched, found_sched
 
 
-    # if there is no local scheduler, return None
-    return None, found_sched
+def get_local_sched(found_sched):
```
Comment (boegel) [inline lib/vsc/mympirun/rm/sched.py]: new function, new dedicated test?

try to also test the return of `None` (which may not be easy)

Also, I'd prefer this:

```python
res = None
for sched in found_sched:
    if sched._is_sched_for('local'):
        res = sched
        break
return res

File: lib/vsc/mympirun/rm/sched.py [hunk 7]
```diff
@@ -42,27 +42,36 @@ def what_sched(requested):
...
     if requested:
         for sched in found_sched:
             if sched._is_sched_for(requested):
                 return sched, found_sched
         LOGGER.warn("%s scheduler was requested, but mympirun failed to find an implementation", requested)
 
+    # if PBS_NODEFILE is not present, use local scheduler
+    if not os.getenv('PBS_NODEFILE'):
+        LOGGER.debug("PBS_NODEFILE not found, using local scheduler")
+        return local_sched, found_sched
```
Comment (boegel) [inline lib/vsc/mympirun/rm/sched.py]: hmm, looking at the logic that is there now, why is this actually needed?

Isn't the loop below supposed to only return a `PBS` scheduler in case we're actually in a PBS job?

It does check for `$PBS_JOBID` rather than `$PBS_NODEFILE` though, but it seems to suggest the actual bug is somewhere else, and we're cheating here?

File: test/end2end.py (hunk 5 above)
Comment (Caylo) [inline test/end2end.py]: the cleanup_PBS_env in tearDown tries to `os.remove(os.environ['PBS_NODEFILE'])`, which fails if PBS_NODEFILE isn't present

Comment (boegel) [inline test/end2end.py]: rather than using `cleanup_PBS_env`, we should just restore the original environment in `tearDown` using

```python
os.environ = self.orig_environ
```

so, let's trash `cleanup_PBS_env`?

File: test/end2end.py [hunk 8]
```diff
@@ -308,3 +311,14 @@ def change_env(self, cores):
...
+
+    def test_unset_nodefile(self):
+        self.assertEqual(schedm.what_sched(False)[0], PBS)
+        nodefile = os.environ['PBS_NODEFILE']
+        del os.environ['PBS_NODEFILE']
+        # fall back to local if PBS_NODEFILE is not available
+        self.assertEqual(schedm.what_sched(False)[0], Local)
+
+        # reset nodefile for proper test cleanup
+        os.environ['PBS_NODEFILE'] = nodefile
```
Comment (boegel) [inline test/end2end.py]: this should not be needed, get rid of `cleanup_PBS_env` instead in tearDown and just restore `self.orig_environ`

**Comment by boegel** [comment]
lgtm, thanks @Caylo!
